### PR TITLE
Linguist repository statistics

### PR DESCRIPTION
**Pull request by mohamed-abdelnour**
Following the discussion in #1677.

### CODE REVIEW

**Comment by Enselic** [comment]
Thanks! LGTM. I will wait some days before merging to give others a chance to chime in.

**Review by Enselic — APPROVED** [review]

**Comment by mohamed-abdelnour** [comment]
Sounds great—thank you!

**Comment by keith-hall** [comment]
I'm not convinced that scripts should be ignored by linguist - we use those to apply patches to syntaxes and themes etc. I agree with hiding the test files etc though, that makes sense.

**Comment by mohamed-abdelnour** [comment]
I also think we shouldn't ignore scripts; however, they are already being ignored in `master`, so I didn't want to make Linguist detect files that are already being ignored. But given that we're having this discussion now let's explore our options!

As of the latest `master`, Linguist detects 196 files inside the `tests` directory:

```console
[bat] [master]
$ github-linguist --breakdown | rg '^tests/.*$' | wc -l
196
```

Of these 196 files, 181 are inside either `tests/syntax-tests/highlighted` or `tests/syntax-tests/source`. I think we agree that these should be ignored.

```console
[bat] [master]
$ github-linguist --breakdown | rg '^tests/syntax-tests/(highlighted|source).*$' | wc -l
181
```

Here are the 15 remaining files:

```console
[bat] [master]
$ github-linguist --breakdown | rg --pcre2 '^tests/(?!syntax-tests/(highlighted|source)).*$'
tests/assets.rs
tests/integration_tests.rs
tests/no_duplicate_extensions.rs
tests/snapshot_tests.rs
tests/tester.rs
tests/utils/mocked_pagers.rs
tests/utils/mod.rs
tests/examples/regression_tests/issue_314.hs
tests/examples/regression_tests/issue_914.rb
tests/examples/regression_tests/issue_915.vue
tests/examples/regression_tests/issue_985.js
tests/mocked-pagers/more
tests/mocked-pagers/most
tests/mocked-pagers/more.bat
tests/mocked-pagers/most.bat
```

If this PR gets merged as is, Linguist would only detect the following 7 files inside the `tests` directory:

```console
[bat] [clean-up-repo-stats]
$ github-linguist --breakdown | rg '^tests/.*$'
tests/assets.rs
tests/integration_tests.rs
tests/no_duplicate_extensions.rs
tests/snapshot_tests.rs
tests/tester.rs
tests/utils/mocked_pagers.rs
tests/utils/mod.rs
```

Lastly, this is what I think is the most sane option (obviously very opinionated): mohamed-abdelnour/bat@bca7179b9e8a37058202c4ec5aba147b96d7d48d. (I added blank lines to make the changes more obvious for now.)

If we apply this patch, Linguist would detect the following 15 files inside the `tests` directory:

```console
[bat] [clean-up-repo-stats-v2]
$ github-linguist --breakdown | rg '^tests/.*$'
tests/assets.rs
tests/integration_tests.rs
tests/no_duplicate_extensions.rs
tests/snapshot_tests.rs
tests/tester.rs
tests/utils/mocked_pagers.rs
tests/utils/mod.rs
tests/benchmarks/comparison.sh
tests/benchmarks/run-benchmarks.sh
tests/syntax-tests/regression_test.sh
tests/syntax-tests/update.sh
tests/scripts/find-slow-to-highlight-files.py
tests/snapshots/generate_snapshots.py
tests/syntax-tests/compare_highlighted_versions.py
tests/syntax-tests/create_highlighted_versions.py
```

Output of running Linguist:

```console
[bat] [clean-up-repo-stats-v2]
$ github-linguist
94.17%  Rust
4.37%   Python
1.45%   Shell
```

### TL;DR

Compared to `master`, merging this PR would make Linguist ignore the following:

- Files inside `tests/syntax-tests/highlighted`
- Files inside `tests/syntax-tests/source`
- `tests/examples/regression_tests/issue_314.hs`
- `tests/examples/regression_tests/issue_914.rb`
- `tests/examples/regression_tests/issue_915.vue`
- `tests/examples/regression_tests/issue_985.js`
- `tests/mocked-pagers/more`
- `tests/mocked-pagers/most`
- `tests/mocked-pagers/more.bat`
- `tests/mocked-pagers/most.bat`

bca7179b9e8a37058202c4ec5aba147b96d7d48d does the same thing but makes Linguist _detect_ these 8 extra scripts:

- `tests/benchmarks/comparison.sh`
- `tests/benchmarks/run-benchmarks.sh`
- `tests/syntax-tests/regression_test.sh`
- `tests/syntax-tests/update.sh`
- `tests/scripts/find-slow-to-highlight-files.py`
- `tests/snapshots/generate_snapshots.py`
- `tests/syntax-tests/compare_highlighted_versions.py`
- `tests/syntax-tests/create_highlighted_versions.py`

Looking forward to getting your feedback!

**Comment by Enselic** [comment]
Warning: I am now entering nit-pick mode.

Looking at bca7179, I have the following comments.

First, I don't think `mocked-pagers` should be ignored, because they are executed as part of integration tests. So IMO they are just as much part of the bat code base as say `tests/integration_tests.rs`.

Secondly, when it comes to the three occurances of `-linguist-vendored`, I actually prefer a simple-to-maintain `.gitattributes` over a 100% accurate one. In other words, I prefer to view these files as casualties of an easy-to-main `.gitattributes`. So IMHO those three lines should be removed, so that there are only directory wildcard lines left to maintain.

Otherwise it looks good!

**Comment by mohamed-abdelnour** [comment]
> Warning: I am now entering nit-pick mode in this PR.

I absolutely love how much effort we're all putting into this! :)

> Lastly, if we end up having just directory wildcard lines, I think it's nicer to not separate them by newlines.

Agreed, I was going to remove them before finalising this. Do we keep the comment, though?

> I do like that you sorted them alphabetically though.

I knew someone would notice; I just had to! :)

**Comment by mohamed-abdelnour** [comment]
So it comes down to one of these 2 options:

Exactly as @Enselic suggested: 70d3ec090c5dba6dc146670a47c03a0d77c7ed53

```console
[bat] [clean-up-repo-stats-v3]
$ github-linguist
95.92%  Rust
3.79%   Python
0.28%   Shell
0.01%   Batchfile
[bat] [clean-up-repo-stats-v3]
$ github-linguist --breakdown | rg '^tests/.*$'
tests/assets.rs
tests/integration_tests.rs
tests/no_duplicate_extensions.rs
tests/snapshot_tests.rs
tests/tester.rs
tests/utils/mocked_pagers.rs
tests/utils/mod.rs
tests/mocked-pagers/more
tests/mocked-pagers/most
tests/syntax-tests/regression_test.sh
tests/syntax-tests/update.sh
tests/mocked-pagers/more.bat
tests/mocked-pagers/most.bat
tests/scripts/find-slow-to-highlight-files.py
tests/syntax-tests/compare_highlighted_versions.py
tests/syntax-tests/create_highlighted_versions.py
```

Or, this attempt to minimise "casualties" (love the description): e00ffff1433ed3c800c3edd50b04a0b4278dc0e2

```console
[bat] [clean-up-repo-stats-v4]
$ github-linguist
94.13%  Rust
4.37%   Python
1.48%   Shell
0.01%   Batchfile
[bat] [clean-up-repo-stats-v4]
$ github-linguist --breakdown | rg '^tests/.*$'
tests/assets.rs
tests/integration_tests.rs
tests/no_duplicate_extensions.rs
tests/snapshot_tests.rs
tests/tester.rs
tests/utils/mocked_pagers.rs
tests/utils/mod.rs
tests/benchmarks/comparison.sh
tests/benchmarks/run-benchmarks.sh
tests/mocked-pagers/more
tests/mocked-pagers/most
tests/syntax-tests/regression_test.sh
tests/syntax-tests/update.sh
tests/mocked-pagers/more.bat
tests/mocked-pagers/most.bat
tests/scripts/find-slow-to-highlight-files.py
tests/snapshots/generate_snapshots.py
tests/syntax-tests/compare_highlighted_versions.py
tests/syntax-tests/create_highlighted_versions.py
```

Please chime in with what you think, and be sure to nitpick the comments! :)

**Comment by Enselic** [comment]
e00ffff is a significant improvement over bca7179 in ease-of-maintenance, but even maintaining file extensions is too much work long term in this case I think. So I prefer 70d3ec0. But if I get outnumbered, I am fine with e00ffff.

Regarding the comment: if I would have to choose, I would choose to keep the comment.

I agree it is a amusing that we are putting this much effort into something like this :) I still appreciate your systematic and thorough way to attack the problem though :+1:

**Comment by mohamed-abdelnour** [comment]
I appreciate your feedback and patience through all this—thank you! I was worried this would be a little silly when I first thought of opening an issue. :)
Let’s wait for the others to chime in, and move forward with whichever implementation ends up being more popular!

**Review by sharkdp — APPROVED** [review]
Looks good to me - thank you!

Edit: I just now saw that I should probably vote for a certain commit :-). 70d3ec0 looks like the easiest for now, so I would prefer that as well.

**Comment by mohamed-abdelnour** [comment]
Thank you!

Well, 2 out of the 4 maintainers chose 70d3ec090c5dba6dc146670a47c03a0d77c7ed53, so it probably makes sense to move forward with it. I'll let you decide how long to wait for the others before merging; but in the mean time, shall I update this PR to 70d3ec090c5dba6dc146670a47c03a0d77c7ed53?

If so, do I just add a second commit, or do I rebase so that only one commit shows up in the history?

**Comment by Enselic** [comment]
If you are comfortable with it, feel free to force push a new commit. Who does not love a neat git commit log? 😍

**Comment by mohamed-abdelnour** [comment]
> Who does not love a neat git commit log? :heart_eyes:

Exactly! :)

**Comment by Enselic** [comment]
@keith-hall Does this look ok to you too now?

**Review by keith-hall — APPROVED** [review]

**Comment by Enselic** [comment]
Ok then! Let's merge this.

**Comment by mohamed-abdelnour** [comment]
Looking good! :)

Before | After
:--- | :---
 |
